src/libostree/ostree-repo-finder-mount.c

```diff
@@ -0,0 +1,497 @@
+ *
+ * Since: 2017.6
+ */
+
+typedef GList/*<owned GObject>*/ ObjectList;
+
+static void
+object_list_free (ObjectList *list)
+{
+  g_list_free_full (list, g_object_unref);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ObjectList, object_list_free)
+
+static void ostree_repo_finder_mount_iface_init (OstreeRepoFinderInterface *iface);
+
+struct _OstreeRepoFinderMount
+{
+  GObject parent_instance;
+
+  GVolumeMonitor *monitor;  /* owned */
+};
+
+G_DEFINE_TYPE_WITH_CODE (OstreeRepoFinderMount, ostree_repo_finder_mount, G_TYPE_OBJECT,
+                         G_IMPLEMENT_INTERFACE (OSTREE_TYPE_REPO_FINDER, ostree_repo_finder_mount_iface_init))
+
+/* Get the child of @parent which is named @child_basename, repeatedly resolving
+ * it to its target if it’s a symlink. This does not resolve symlinks in
+ * @parent’s path. If more than 15 symlinks are chained,
+ * %G_IO_ERROR_TOO_MANY_LINKS will be returned. If @parent/@child_basename is
+ * not a symlink, it will be returned immediately. */
+static GFile *
+expand_symlink (GFile         *parent,
+                const gchar   *child_basename,
+                GCancellable  *cancellable,
+                GError       **error)
+{
+  g_autoptr(GFile) resolved_file = NULL;
+  g_autofree gchar *parent_path = NULL;
+  guint i;
+  const guint n_attempts = 15;  /* arbitrarily chosen */
+
+  resolved_file = g_file_get_child (parent, child_basename);
+
+  for (i = 0; i < n_attempts; i++)
+    {
+      g_autoptr(GError) local_error = NULL;
+      g_autoptr(GFileInfo) file_info = NULL;
+
+      file_info = g_file_query_info (resolved_file,
+                                     G_FILE_ATTRIBUTE_STANDARD_TYPE ","
+                                     G_FILE_ATTRIBUTE_STANDARD_SYMLINK_TARGET,
+                                     G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                     cancellable,
+                                     &local_error);
+
+      if (local_error != NULL)
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      if (g_file_info_get_file_type (file_info) == G_FILE_TYPE_SYMBOLIC_LINK)
+        {
+          g_set_object (&resolved_file,
+                        g_file_resolve_relative_path (parent,
+                                                      g_file_info_get_symlink_target (file_info)));
+          continue;
+        }
+      else
+        {
+          return g_steal_pointer (&resolved_file);
+        }
+    }
+
+  /* Reached the retry limit. */
+  parent_path = g_file_get_path (parent);
+  g_set_error (error, G_IO_ERROR, G_IO_ERROR_TOO_MANY_LINKS,
+               "Too many symlink links when resolving ‘%s/%s’.",
+               parent_path, child_basename);
+
+  return NULL;
+}
+
+/* Resolve all symlinks in @file’s path, returning a normalised, canonical path
+ * containing no symlinks. This returns the same errors as expand_symlink(). */
+static GFile *
+expand_all_symlinks (GFile         *file,
```

Comment:
Yes, it’s effectively a version of `realpath()` for `GFile`s. Using `GFile` here is a bit pointless, since the file system has to be local in order to have an OSTree repository on it, so `GFile` brings no benefits. I’ll try and find time to rewrite this to use FDs.